### PR TITLE
Allow only [a-z0-9] at the start of package name

### PR DESCRIPTION
**Pull request by ghostdevhv**
### Change
- Added a condition to not allow the first character to be a special/uppercase character. 
- Fixed and added the test cases. 
  
cc: @nishantms 

See: github/npm#2142

### CODE REVIEW

File: README.md
Comment (amit-avit) [inline README.md:26]:
```suggestion
- package name should not start with a special character
```

File: README.md
Comment (amit-avit) [inline README.md:27]: remove full stop for consistency

```suggestion
- package name should not have any uppercase letter
```

File: index.js
Comment (amit-avit) [inline index.js:42]: instead of `capital character`, either use `uppercase character` or `uppercase letter` or `capital letter` t/o

```suggestion
    warnings.push('name cannot start with a special or uppercase character')
```

File: test/index.js
Comment (amit-avit) [inline test/index.js:25]:
```suggestion
    warnings: ['name cannot start with a special or uppercase character']
```

File: test/index.js
Comment (amit-avit) [inline test/index.js:22]: Scoped public packages can start with @ so isn't this testcase incorrect ?

File: README.md
Comment (Soph) [inline README.md:27]:
```suggestion
- all the characters in the package name must be lowercase i.e., no uppercase or mixed case names are allowed
```

Keep the original line?

**Review by Soph — APPROVED** [review]

File: test/index.js
Comment (amit-avit) [inline test/index.js:28]: We should also have a test case which checks for a special character in the middle instead of start.

File: test/index.js
Comment (amit-avit) [inline test/index.js:115]: Previous message seems to be more appropriate here.

File: test/index.js
Comment (amit-avit) [inline test/index.js:68]: This test case is testing for start with whitespace, so this messaging seems inaccurate.

File: test/index.js
Comment (amit-avit) [inline test/index.js:69]: Lets add another test case of a trailing whitespace if not already present.

File: index.js
Comment (amit-avit) [inline index.js:39]: With your new conditional check we can get rid of the checks between line# 33 and 39

File: index.js
Comment (amit-avit) [inline index.js:41]: Should @ be allowed as 1st char as scoped package names needs that ?

**Review by amit-avit — CHANGES_REQUESTED** [review]

File: README.md
Comment (amit-avit) [inline README.md:32]: Update the examples section below.

File: README.md
Comment (ghostdevhv) [reply README.md]: Done

File: index.js
Comment (ghostdevhv) [reply index.js:39]: Actually no. Because for the old packages, we should throw the error only when the package name is starting with `period` and `underscore`. If we remove those changes, it will start throwing the error for all the special characters.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js:28]: So, there is a condition that the package name should not contain `~)('!*` these characters.  Apart from these and `non-url-safe` (refer to the readme for this) characters, all the characters are allowed and we do have test cases to validate these conditions. 

I don' think we need another similar kind of test case. Let me know if I'm missing anything.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js:69]: We do have a test case for that condition.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js]: So, what happens is each package name gets checked against all the rules created in the `index.js`, and errors and warnings got pushed in the list so while writing a test case, we have to mention all the warnings. That' why there are 2 warnings here.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js:22]: Added a message in the issue and tagged Stefan to confirm this.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js]: So, what happens is each package name gets checked against all the rules created in the index.js, and errors and warnings got pushed in the list so while writing a test case, we have to mention all the warnings. That' why there are 2 warnings here.

File: index.js
Comment (ghostdevhv) [reply index.js:41]: Added a comment in the issue and tagged Stefan to confirm this

File: README.md
Comment (ghostdevhv) [reply README.md:32]: Will do it once I get the confirmation from Stefan regarding the "@" at the beginning of package name.

File: test/index.js
Comment (amit-avit) [reply test/index.js:28]: Based on the new check flow would return from there as it starts with @ and the other condition that `!` is allowed in the middle of name would not get validated.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js:22]: Corrected it. @ should be allowed

**Comment by ghostdevhv** [comment]
@Soph -  Made all the required changes. Tested it on stage env too. 
Do we need to change the package version in package.json too (npm/validate-npm-package-name#33)?
Since this is a package, what's the process to push our changes in prod? Do we have any doc for that?

cc: @amit-avit @nishantms

File: README.md
Comment (ghostdevhv) [reply README.md:32]: No need to update the examples now.

File: index.js
Comment (ghostdevhv) [reply index.js]: Yes, it should be allowed. Fixed it.

**Review by nishantms — APPROVED** [review]

**Review by Soph — APPROVED** [review]

File: test/index.js
Comment (amit-avit) [inline test/index.js:26]: `@npm` is an invalid name, a test could be added for this.

**Review by amit-avit — APPROVED** [review]

File: index.js
Comment (ghostdevhv) [reply index.js:42]: Because this condition is only valid for new packages. If we convert this into an error, then we will get error while publishing older packages that do not support this condition.

File: test/index.js
Comment (ghostdevhv) [reply test/index.js:26]: Umm, that's a loophole in the code. Ideally, the code should have some sort of function that differentiates between scope name and package name but right now it considers everything as a package name. 

But if we try to publish the package name starting with "@", it will get failed before hitting this function (In user-acl-two service).

Comment (Soph) [reply test/index.js:26]: I wonder if we should move this check further down where also scopes are checked and then only apply it to the actual package name and not the scope. That should solve the `@` issue and make it less risky? 

https://github.com/npm/validate-npm-package-name/blob/main/index.js#L76-L88

Comment (Soph) [reply test/index.js:26]: But I see, it's only about `@npm` being blocked in `user-acl-two`

Comment (ghostdevhv) [reply test/index.js:26]: I've added the test case for this condition.

File: index.js
Comment (assapir) [reply index.js:42]: won't it be explicit in line 69 (72)? why this change is needed?

Comment (ghostdevhv) [reply index.js:42]: Initially, we allowed the special characters (except dot and underscore)  as a first letter for the package name. e.g. there is a package name `-` exists. 
Now, the package name should only start with the lowercase character and the digit. 

Overall the validation happens in two places. one is on `user-acl-two` side and other is in this package.

File: package.json
Comment (wraithgar) [inline package.json:3]: This change should be made during the release process and not as part of the PR

Comment (Soph) [reply package.json:3]: @ghostdevhv sorry, I haven't really explained this well enough. 

The idea is that the PR is approved and merged and then after merging you update package.json (or someone), publish the package and then just commit the version number change directly to main.

Comment (nishantms) [reply package.json:3]: @Soph : Sure will revert this changes.

**Comment by wraithgar** [comment]
Closing this for now, as it looks like we aren't going to be able to land this here because of the need to support multiple registries.

This package prevents folks from trying to publish packages with broadly known invalid names i.e. builtins.  It can't also do things like include common spam terms, or things only certain registries implement.  At the end of the day it's a client validation library meaning it's a ux helper, not a security tool.

**Pull request closed without merging**
